components/camel-solr/src/main/java/org/apache/camel/component/solr/SolrComponent.java

```diff
@@ -86,38 +113,105 @@ public SolrComponent() {
+        } else {
+            solrClientReference = solrClientMap.get(signature);
+        }
+        // register producer against solrClient (for later close of client)
+        solrClientReference.registerSolrProducer(solrProducer);
+        return solrClientReference.getSolrClient();
+    }
+
+    private void addSolrClientToSolrServerReference(SolrConfiguration solrConfiguration, SolrClient solrClient) {
+        SolrEndpoint solrEndpoint = solrConfiguration.getSolrEndpoint();
+        if (solrEndpoint != null) {
+            SolrServerReference solrServerReference = servers.get(solrEndpoint);
+            if (solrServerReference == null) {
+                solrServerReference = new SolrServerReference();
+                servers.put(solrEndpoint, solrServerReference);
+            }
+            if (solrClient instanceof CloudSolrClient) {
+                solrServerReference.setCloudSolrServer((CloudSolrClient) solrClient);
+            }
+            if (solrClient instanceof ConcurrentUpdateSolrClient) {
+                solrServerReference.setUpdateSolrServer((ConcurrentUpdateSolrClient) solrClient);
+            }
+            if (solrClient instanceof HttpSolrClient) {
+                solrServerReference.setSolrServer((HttpSolrClient) solrClient);
+            }
+        }
+    }
+
+    public void closeSolrClient(SolrProducer solrProducer) {
+        // close when generated for endpoint
+        List<String> signatureToRemoveList = new ArrayList<>();
+        for (Map.Entry<String, SolrClientReference> entry : solrClientMap.entrySet()) {
+            SolrClientReference solrClientReference = entry.getValue();
+            if (solrClientReference.unRegisterSolrProducer(solrProducer) == 0) {
+                signatureToRemoveList.add(entry.getKey());
+            }
+        }
+        removeFromSolrClientMap(signatureToRemoveList);
+    }
+
+    private void removeFromSolrClientMap(Collection<String> signatureToRemoveList) {
+        for (String signature : signatureToRemoveList) {
+            SolrClientReference solrClientReference = solrClientMap.get(signature);
+            solrClientMap.remove(signature);
+            try {
+                solrClientReference.getSolrClient().close();
+            } catch (IOException e) {
+                LOG.warn("Error shutting down solr client. This exception is ignored.", e);
+            }
+        }
+    }
+
+    @Deprecated
     public SolrServerReference getSolrServers(SolrEndpoint endpoint) {
         return servers.get(endpoint);
     }
 
+    @Deprecated
     public void addSolrServers(SolrEndpoint endpoint, SolrServerReference servers) {
         this.servers.put(endpoint, servers);
     }
 
     @Override
     protected void doShutdown() throws Exception {
+        removeFromSolrClientMap(solrClientMap.keySet());
         for (SolrServerReference server : servers.values()) {
             shutdownServers(server);
         }
         servers.clear();
     }
 
+    @Deprecated
```

Comment:
If those methods are not used anymore, you can safely remove them. We are going to release an LTS release with 3.11.0, so we are under heavy developments, so you can remove methods like these.

Reply:
ok, thanks